### PR TITLE
Add build_workflow library

### PR DESCRIPTION
**Pull request by camerski**
### Description
Adds `Builder`, `BuildRecorder`, and `GitRepository` classes that represent the build process for a component. This is a refactoring of the old classes in `lib` that simplifies and cleanly separates logic and responsibilities.

### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/157
 
### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by mch2** [comment]
Can we combine this with #168?  I think its easier to digest as a single PR

File: bundle-workflow/python/build_workflow/builder.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Copyright OpenSearch Contributors.
```
Comment (dblock) [inline bundle-workflow/python/build_workflow/builder.py:1]: This looks like a variation/reimplementation of https://github.com/opensearch-project/opensearch-build/blob/main/bundle-workflow/python/lib/component.py without dataclass.

File: bundle-workflow/python/build_workflow/builder.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        if build_script is None:
+            raise RuntimeError(f'Could not find build.sh script. Looked in {paths}')
+
+        build_command = f'{build_script} {version} {arch}'
+        self.git_repo.execute(build_command)
+        self.output.record_component(self.component_name, self.git_repo)
+
+    def export_artifacts(self):
+        artifacts_dir = os.path.realpath(os.path.join(self.git_repo.dir.name, 'artifacts'))
+        for artifact_type in ["maven", "plugins", "libs"]:
```
Comment (mch2) [inline bundle-workflow/python/build_workflow/builder.py:43]: this is missing "bundle".  Opensearch core's build script will place the min bundle in that directory.

Comment (camerski) [reply bundle-workflow/python/build_workflow/builder.py:43]: Ah, I missed that. Good catch!

**Comment by camerski** [comment]
https://github.com/opensearch-project/opensearch-build/pull/170

**Pull request closed without merging**